File: 0189-rotate-array/0189-rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k = k % n;
        if (k == 0)
            return;
        int lasthalf = n - k;

        int[] ans = new int[nums.length];
        int idx = 0;

        for (int i = lasthalf; i < n; i++) {
            ans[idx++] = nums[i];
        }

        for (int i = 0; i < lasthalf; i++) {
            ans[idx++] = nums[i];
        }

        for (int i = 0; i < nums.length; i++) {
            nums[i] = ans[i];
        }

    }
}

File: 0189-rotate-array/0189-rotate-array.py
from typing import List

class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        n = len(nums)
        k = k % n 
        if k == 0:
            return
        lasthalf = n - k

        ans = [0] * n
        idx = 0

        for i in range(lasthalf, n):
            ans[idx] = nums[i]
            idx += 1

        for i in range(0, lasthalf):
            ans[idx] = nums[i]
            idx += 1

        for i in range(n):
            nums[i] = ans[i]
